src/core/net/ip6.cpp

```diff
@@ -603,10 +618,243 @@ otError Ip6::HandleOptions(Message &aMessage, Header &aHeader, bool &aForward)
     return error;
 }
 
-otError Ip6::HandleFragment(Message &aMessage)
+#if OPENTHREAD_ENABLE_IP6_FRAGMENTATION
+otError Ip6::HandleFragmentation(Message &aMessage, IpProto aIpProto)
+{
+    otError        error = OT_ERROR_NONE;
+    Header         header;
+    FragmentHeader fragmentHeader;
+    Message *      fragment        = NULL;
+    uint16_t       fragmentCnt     = 0;
+    uint16_t       payloadFragment = 0;
+    uint16_t       offset          = 0;
+    uint16_t maxPayloadFragment = MAKE_DIVIDABLE_BY_EIGHT(kMinimalMtu - aMessage.GetOffset() - sizeof(fragmentHeader));
+    uint16_t payloadLeft        = aMessage.GetLength() - aMessage.GetOffset();
+
+    if (aMessage.GetLength() > kMaxAssembledDatagramLength)
+    {
+        ExitNow(error = OT_ERROR_NO_BUFS);
+    }
+
+    VerifyOrExit(aMessage.Read(0, sizeof(header), &header) == sizeof(header), error = OT_ERROR_PARSE);
+    header.SetNextHeader(kProtoFragment);
+
+    fragmentHeader.Init();
+    fragmentHeader.SetIdentification(Random::NonCrypto::GetUint32());
+    fragmentHeader.SetNextHeader(aIpProto);
+    fragmentHeader.SetMoreFlag();
+
+    while (payloadLeft != 0)
+    {
+        if (payloadLeft < maxPayloadFragment)
+        {
+            fragmentHeader.ClearMoreFlag();
+            payloadFragment = payloadLeft;
+            payloadLeft     = 0;
+            otLogDebgIp6("Last Fragment");
+        }
+        else
+        {
+            payloadLeft -= maxPayloadFragment;
+            payloadFragment = maxPayloadFragment;
+        }
+        offset = fragmentCnt * (maxPayloadFragment >> 3);
+        fragmentHeader.SetOffset(offset);
+        VerifyOrExit((fragment = NewMessage(0)) != NULL, error = OT_ERROR_NO_BUFS);
+        SuccessOrExit(error = fragment->SetLength(aMessage.GetOffset() + sizeof(fragmentHeader) + payloadFragment));
+        header.SetPayloadLength(payloadFragment + sizeof(fragmentHeader));
+        assert(fragment->Write(0, sizeof(header), &header) == sizeof(header));
```

Comment:
Avoid putting logic inside `assert()` - `assert()` may be compiled out depending on the build configuration.

Reply:
Yeah, my mistake. Thank you for your remark.